python/ray/tests/kuberay/scripts/scale_up.py

```diff
@@ -1,9 +1,16 @@
 import ray
+import time
 
 
 def main():
     """Submits CPU request."""
     ray.autoscaler.sdk.request_resources(num_cpus=2)
+    # Hold the driver for a bit to allow the message
+    # "Adding 1 nodes of type small-group" to reach the driver's stdout.
+    time.sleep(15)
```

Comment:
Instead of doing this, can we use `test_utils.wait_for_condition`?

Reply:
I think this is all doable, but probably not worth the trouble -- not receiving the logs in 15 seconds is definitely a test failure.

Reply:
Let me add some more comments explaining the intent of the test.

Reply:
2 options

1. The script could wait for the node to come online and wait on `ray.nodes()` or `ray.available_resources()["CPU"]` or something.
2. You could use `test_utils.get_log_batch()` to get the log batch in this test case.

I'd advocate for option 2.

Reply:
Option 1 doesn't quite work since node scale-up could happen before this message is issued.
Let me take a look at `test_utils.get_log_batch()`.

Reply:
Issues with using `test_utils.get_log_batch`:

1. That would test that some Ray-internal machinery has processed the logs. I want to verify e2e that the logs got to stdout.
2. I don't want to spend time understanding this method, the method's input and outputs, and the Ray-internal machinery involved in processing the logs.
3. We'd be polling with a timeout of 15 seconds to see if the relevant logs have been processed... which is the same thing as waiting for 15 seconds and then checking if the logs are there.

Reply:
Another option would be to have this script sleep forever, run the script using subprocess.Popen, and poll output with Popen.communicate with a time limit of 15 seconds.
...But see item 3. above.

Reply:
> Option 1 doesn't quite work since node scale-up could happen before this message is issued.

I don't think this is possible. We always print the message before we launch new nodes (since node launch can take a lot of time).

Reply:
> I don't think this is possible. We always print the message before we launch new nodes (since node launch can take a lot of time).

That's not quite true. The event summarizer class processes the node launch event before node launch, but the message is logged after the autoscaler update and emitted to logs after that. Given that we're in a KinD environment with pre-loaded images, it is feasible but unlikely that the Ray worker will start up quickly before the logs are emitted.

> btw with wait_for_condition i think you want to keep the retry interval small (probably just the default) and set the timeout high (leave it at 15s), that way it has the opportunity to finish quickly.

This would shave about 10 seconds off of the test's runtime. The test takes currently takes at least 40min, most of which is image build.

Reply:
I can wrap the existing script in subprocess.Popen or similar.

Reply:
Ok, that wasn't so bad.
Modified to redirect stdout to an in-memory object and use wait_for_condition.